pkg/vm/engine/tae/containers/cn_vector_adapter.go

```diff
@@ -0,0 +1,604 @@
+	slice := vec.Slice().([]T)
+	return unsafe.Pointer(&slice[0])
+}
+
+func (vec *CnTaeVector[T]) AppendNoNulls(s any) {
+	panic("Soon Deprecated")
+}
+
+// TODO: --- We can remove below function as they are only used in Testcases.
+
+func (vec *CnTaeVector[T]) Delete(delRowId int) {
+	deletes := roaring.BitmapOf(uint32(delRowId))
+	vec.Compact(deletes)
+}
+
+func (vec *CnTaeVector[T]) ReadFromFile(f common.IVFile, buffer *bytes.Buffer) (err error) {
+	stat := f.Stat()
+	var n []byte
+	var buf []byte
+	var tmpNode []byte
+	if stat.CompressAlgo() != compress.None {
+		osize := int(stat.OriginSize())
+		size := stat.Size()
+		tmpNode, err = vec.GetAllocator().Alloc(int(size))
+		if err != nil {
+			return
+		}
+		defer vec.GetAllocator().Free(tmpNode)
+		srcBuf := tmpNode[:size]
+		if _, err = f.Read(srcBuf); err != nil {
+			return
+		}
+		if buffer == nil {
+			n, err = vec.GetAllocator().Alloc(osize)
+			if err != nil {
+				return
+			}
+			buf = n[:osize]
+		} else {
+			buffer.Reset()
+			if osize > buffer.Cap() {
+				buffer.Grow(osize)
+			}
+			buf = buffer.Bytes()[:osize]
+		}
+		if _, err = compress.Decompress(srcBuf, buf, compress.Lz4); err != nil {
+			if n != nil {
+				vec.GetAllocator().Free(n)
+			}
+			return nil
+		}
+	}
+
+	_, err = vec.ReadFrom(bytes.NewBuffer(buf))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// TODO: --- Below Functions Can be implemented in CN Vector.
+
+func (vec *CnTaeVector[T]) Equals(o Vector) bool {
+
+	if vec.Length() != o.Length() {
+		return false
+	}
+	if vec.GetType() != o.GetType() {
+		return false
+	}
+	if vec.Nullable() != o.Nullable() {
+		return false
+	}
+	if vec.HasNull() != o.HasNull() {
+		return false
+	}
+	if vec.HasNull() {
+		if !vec.NullMask().Equals(o.NullMask()) {
+			return false
+		}
+	}
+	mask := vec.NullMask()
+	for i := 0; i < vec.Length(); i++ {
+		if mask != nil && mask.ContainsInt(i) {
+			continue
+		}
+		var v T
+		if _, ok := any(v).([]byte); ok {
+			if !bytes.Equal(vec.Get(i).([]byte), o.Get(i).([]byte)) {
+				return false
+			}
+		} else if _, ok := any(v).(types.Decimal64); ok {
+			d := vec.Get(i).(types.Decimal64)
+			od := vec.Get(i).(types.Decimal64)
+			if d.Ne(od) {
+				return false
+			}
+		} else if _, ok := any(v).(types.Decimal128); ok {
+			d := vec.Get(i).(types.Decimal128)
+			od := vec.Get(i).(types.Decimal128)
+			if d.Ne(od) {
+				return false
+			}
+		} else if _, ok := any(v).(types.TS); ok {
+			d := vec.Get(i).(types.TS)
+			od := vec.Get(i).(types.TS)
+			if types.CompareTSTSAligned(d, od) != 0 {
+				return false
+			}
+		} else if _, ok := any(v).(types.Rowid); ok {
+			d := vec.Get(i).(types.Rowid)
+			od := vec.Get(i).(types.Rowid)
+			if types.CompareRowidRowidAligned(d, od) != 0 {
+				return false
+			}
+		} else {
+			if vec.Get(i) != o.Get(i) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (vec *CnTaeVector[T]) ForeachWindow(offset, length int, op ItOp, sels *roaring.Bitmap) (err error) {
```

Comment:
Are we really using this function?   What does it do?   Can we remove it?    I would like to get rid of roaring in all places.

Reply:
ForeachWindow is used in `TAE index`
- [Tree](https://github.com/arjunsk/matrixone/blob/e5e451c6bee40b5c116bdd5e699dadb29d07059d/pkg/vm/engine/tae/index/tree.go#L93)
- [ZoneMap](https://github.com/arjunsk/matrixone/blob/e5e451c6bee40b5c116bdd5e699dadb29d07059d/pkg/vm/engine/tae/index/zonemap.go#L142)
- [Multi Index](https://github.com/arjunsk/matrixone/blob/e5e451c6bee40b5c116bdd5e699dadb29d07059d/pkg/vm/engine/tae/tables/indexwrapper/mutindex.go#L124)

Yes, we can migrate the RoaringMap to Normal BitMap once, the adapter passes the UT and BVT.